lahja/endpoint.py

```diff
@@ -61,36 +65,177 @@ def from_name(cls, name: str, base_path: Optional[pathlib.Path]=None) -> 'Connec
+    An Endpoint might have several `InboundConnection`s, each of them represents an
+    Endpoint which has connected to the given Endpoint and is attempting to send it
+    messages.
     """
+    def __init__(self, conn: Connection, new_msg_func: Callable[[Broadcast], None]) -> None:
+        self.conn = conn
+        self.new_msg_func = new_msg_func
+
+        self.logger = logging.getLogger('lahja.endpoint.InboundConnection')
+
+        self._notify_lock = asyncio.Lock()
+        self._received_response = asyncio.Condition()
+
+    async def run(self) -> None:
+        try:
+            while self.conn.more_to_read:
+                message = await self.conn.read_message()
+
+                if isinstance(message, Broadcast):
+                    self.new_msg_func(message)
+                elif isinstance(message, SubscriptionsAck):
+                    async with self._received_response:
+                        self._received_response.notify_all()
+                else:
+                    self.logger.error(f'received unexpected message: {message}')
+        except asyncio.IncompleteReadError:
+            # the remote endpoint has disconnected
+            pass
+
+    async def notify_subscriptions_updated(self,
+                                           subscriptions: Set[Type[BaseEvent]],
+                                           block: bool = True) -> None:
+        """
+        Alert the Endpoint which has connected to us that our subscription set has
+        changed. If `block` is True then this function will block until the remote
+        endpoint has acknowledged the new subscription set. If `block` is False then this
+        function will return immediately after the send finishes.
+        """
+        async with self._notify_lock:
+            # The lock ensures only one coroutine can notify this endpoint at any one time
+            # and that no replies are accidentally received by the wrong coroutines.
+            async with self._received_response:
+                await self.conn.send_message(SubscriptionsUpdated(subscriptions, block))
+                if block:
+                    await self._received_response.wait()
+
+
+class OutboundConnection:
+    """
+    Represents a remote Endpoint (called {name}) which has been connected to. The remote
+    endpoint occasionally sends message "backwards" to the endpoint which connected to it.
+    Those messages (SubscriptionsUpdated) specify which kinds of messages the remote
+    Endpoint is subscribed to. All other messages should be dropped.
+    """
+    def __init__(self, name: str, conn: Connection) -> None:
+        self.conn = conn
+        self.name = name
+        self.subscribed_messages: Set[Type[BaseEvent]] = set()
+
+        self.logger = logging.getLogger('lahja.endpoint.OutboundConnection')
+        self._received_subscription = asyncio.Condition()
+
+    async def run(self) -> None:
+        try:
```

Comment:
Same here with reducing the scope of the `try` block to only contain the part that can throw the exception.